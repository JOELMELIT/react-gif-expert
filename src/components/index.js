
// Para ordener alfabeticamente en Visual Studio - ctrl+p -> >as -> Seleccionar Sort Lines Ascending


export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';










