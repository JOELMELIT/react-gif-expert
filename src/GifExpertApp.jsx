
import React, { useState } from 'react'
//import { AddCategory } from './components/AddCategory';
//import { GifGrid } from './components/GifGrid';
import { AddCategory, GifGrid } from './components'; // Archivo de Barril - Toma las referencias del index.js contenida en la carpeta components


export const GifExpertApp = () => {

    const [categories, setCategories] = useState([]);

    const onAddCategory = ( newCategory ) => {

        if(categories.includes(newCategory)){
            return;
        }

        console.log(newCategory);
        setCategories([newCategory, ...categories]);

    }

    return (
        <>
            <h1>GifExpertApp</h1>

            <AddCategory 
                onNewCategory = { (value) => onAddCategory(value)}
            />
            { 
                categories.map( category =>
                    (
                        <GifGrid 
                            key={ category } 
                            categoryy={ category }
                        />
                    )
                ) 
            }   
        </>
    )
}




//////  CODIGO DE PRUEBAS INICIALES ///////

/*
import React, { useState } from 'react'
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid';

export const GifExpertApp = () => {

    // No se recomienda poner el Hook dentro de una conficional 'if' ya que React perderia la referencia del state
    const [categories, setCategories] = useState([ 'One Punch', 'Dragon Ball Z' ]);

    const onAddCategory = ( newCategory ) => {

        if(categories.includes(newCategory)){
            return;
        }

        // La forma en que lo realice
        //const newCategory = ['Videojuegos'];
        //const addingCategory = [...categories, newCategory]
        //setCategories( addingCategory );
        //categories.push('Videojuegos'); No usar el push ya que muta el arreglo, muta el estado

        // Otra forma de agregarlo e insertarlo al inicio
        //setCategories(['FreeFire', ...categories]);

        console.log(newCategory);
        setCategories([newCategory, ...categories]);

        // Otra forma de agregarlo
        //setCategories( cat => [...cat, 'Valorant']); // El callback que se puede enviar a la funcion del useState


    }

    return (
        <>
*/        
//            {/*  titulo */}
//           <h1>GifExpertApp</h1>

//            {/*  Input */}
//            <AddCategory 
                //setCategoriess={ setCategories } // Utilizando props
                
                // Utilizando una mejor forma de la propiedad
//                onNewCategory = { (value) => onAddCategory(value)}
            
//            />

//            {/* Listado de Gif */}
//            {/*<button onClick={ onAddCategory }>Agregar</button>  Codigo de pruebas iniciales*/}
            
/*            <ol>
                { 
                    categories.map( category =>
                        //return <li key={ category }>{ category }</li> // Codigo de pruebas iniciales
                        (
                            <div key={ category }>
                                <h3>{ category }</h3>
                                <li>{ category }</li>
                            </div>  Todo esto se puede sustituir por el componente de GifGrid

                            <GifGrid 
                                key={ category } 
                                categoryy={ category }
                            />
                        )
                    ) 
                } la llave al inicio de categories indica que se utilizara un expression de JS 

            </ol>
*/            
//                {/* Gif Item */}


//        </>
//    )
//}




/* 

ol - order list
:: marker - Lo crea React para saber que ese es el elmento



Respecto a categories.map, no es recomendable usar el indice del map para la key del <li 
Porque React utiliza este este valor de la llave para saber cuando un elemento se eliminó.
Si nosotros por ejemplo borramos la posicion 0 de la lista, siempre sigue existiendo la posición 0 en nuestro arreglo, por lo cual podría cometer un error y eliminar visualmente del arreglo de elementos.
Y podríamos tener una una información que no es la que nosotros estamos esperando.

*/










