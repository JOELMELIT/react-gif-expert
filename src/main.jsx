import React from 'react';
import ReactDOM from 'react-dom/client';
import { GifExpertApp } from './GifExpertApp';
import './styles.css';

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <GifExpertApp />
  </React.StrictMode>,
)



/*
ero qué es un Hook?
Los Hooks son funciones que te permiten “enganchar” el estado de React y el ciclo de vida desde componentes de función.
Los hooks no funcionan dentro de las clases — te permiten usar React sin clases. 

*/

/*
Con el lanzamiento de React 16.8, se introdujeron los hooks como una forma de permitir que los desarrolladores de React 
puedan reutilizar la lógica de estado y efectos secundarios sin necesidad de usar clases. Los hooks ofrecen una forma más 
sencilla y elegante de manejar el estado y los efectos secundarios en React.

La importancia de los hooks en React radica en que permiten a los desarrolladores escribir código más limpio y legible.
En lugar de tener que preocuparse por las complejidades de las clases y los componentes de orden superior, 
los hooks ofrecen una forma más intuitiva y fácil de manejar el estado y los efectos secundarios en los componentes de React.

Además, los hooks también permiten a los desarrolladores crear componentes más modulares y reutilizables, 
lo que hace que el código sea más escalable y fácil de mantener. Esto es especialmente importante para proyectos grandes 
y complejos en los que el mantenimiento del código es crucial para el éxito del proyecto.

En resumen, los hooks en React son una herramienta fundamental para el desarrollo web moderno. 
Proporcionan una forma más fácil y elegante de manejar el estado y los efectos secundarios en los componentes de React, 
lo que permite a los desarrolladores escribir código más limpio, modular y escalable. 
En los próximos puntos, profundizaremos más en lo que son los hooks en React y cómo se utilizan.


*/