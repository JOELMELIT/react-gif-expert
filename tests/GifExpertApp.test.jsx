
import { fireEvent, render, screen } from '@testing-library/react';
import { GifExpertApp } from '../src/GifExpertApp';




describe('Pruebas en el componente <GifExpertApp />', () => { 

    test('test: debe hacer match con el snapshot', () => { 

        render(<GifExpertApp />);
        const { container } = render(<GifExpertApp />);
        expect( container ).toMatchSnapshot();

    });

    const newCategory = 'smartphone';

    //Primero Debe agregar categorias nuevas:
    test('should Add new categories', () => {

        render(<GifExpertApp />);
        const input = screen.getByRole('textbox'); 
        const form = screen.getByRole('form'); 

        //ahora disparo los eventos para agregar 3 categorias nuevas
        fireEvent.input(input, { target: { value: newCategory + ' 1'} });
        fireEvent.submit(form); //agrega smartphones

        fireEvent.input(input, { target: { value: newCategory + ' 2' } });
        fireEvent.submit(form); //agrega smartphones2

        fireEvent.input(input, { target: { value: newCategory + ' 3' } });
        fireEvent.submit(form);//agrega smartphones3
        screen.debug();
        //Espero que se hallan agregado 3 categor√≠as cuyos nombres aparecen en etiquetas h3
        expect(screen.getAllByRole('heading', { level: 3 }).length).toBe(3);
    });


});