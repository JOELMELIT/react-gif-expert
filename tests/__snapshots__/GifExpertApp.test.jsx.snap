// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <GifExpertApp /> test: debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    GifExpertApp
  </h1>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
